pennylane/gradients/gradient_transform.py

```diff
@@ -0,0 +1,178 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains utilities for defining custom gradient transforms,
+including a decorator for specifying gradient expansions."""
+# pylint: disable=too-few-public-methods
+import pennylane as qml
+
+
+unsupported_op = lambda op: op.grad_method is None
+supported_op = lambda op: op.grad_method is not None
+trainable_op = lambda op: any(qml.math.requires_grad(p) for p in op.parameters)
+
+
+def gradient_expand(tape, depth=10):
+    """Expand out a tape so that it supports differentiation
+    of requested operations.
+
+    This is achieved by decomposing all trainable operations that have
+    ``Operation.grad_method=None`` until all resulting operations
+    have a defined gradient method.

```

Comment:
Based on the existing set of operations in PL, will this _always_ be possible?

Reply:
```suggestion
    have a defined gradient method, up to maximum depth ``depth``. Note that this
    might not be possible, in which case the gradient rule will fail to apply.
```